cpp/src/io/utilities/file_io_utilities.hpp

```diff
@@ -124,7 +130,10 @@ class cufile_input : public cufile_io_base {
   virtual std::future<size_t> read_async(size_t offset,
                                          size_t size,
                                          uint8_t* dst,
-                                         rmm::cuda_stream_view stream) = 0;
+                                         rmm::cuda_stream_view stream)
+  {
+    CUDF_FAIL("these classes that support read_async must override it.");
```

Comment:
The default implementation here should use a `read` function to return `std::future` (whose result is immediately available), so derived `datasource` classes have a default working implementation. @devavret and I can help with the implementation, if needed.